libraries/Microsoft.Bot.Builder.Dialogs/ObjectPath.cs

```diff
@@ -256,6 +256,66 @@ public static void ForEachProperty(object obj, Action<string, object> action)
             */
         }
 
+        /// <summary>
+        /// Get all properties in an object.
+        /// </summary>
+        /// <param name="obj">Object to enumerate property names.</param>
+        /// <returns>enumeration of property names on the object if it is not a value type.</returns>
+        public static IEnumerable<string> GetProperties(object obj)
+        {
+            if (obj == null)
+            {
+                yield break;
+            }
+            else if (obj is IDictionary<string, object> dict)
+            {
+                foreach (var entry in dict)
+                {
+                    yield return entry.Key;
+                }
+            }
+            else if (obj is JObject jobj)
+            {
+                foreach (var property in jobj.Properties())
+                {
+                    yield return property.Name;
+                }
+            }
+            else
+            {
+                foreach (var property in obj.GetType().GetProperties().Select(p => p.Name))
```

Comment:
Consider passing binding flags to `GetProperties`, such as `obj.GetType().GetProperties(BindingFlags.Public|BindingFlags.Instance)`. The current code would not return private, protected or internal properties which is good, but would return public static properties. To remediate this,you need BindingFlags.Instance, but once you pass a binding flag, this overload will return all visibility levels unless you specify, which is why you want is public + instance level properties.